### ISSUE TITLE
IoU metric

### ISSUE THREAD

**Issue by ajuric** [qubvel/segmentation_models#68]
In the docs of [`iou_score`](https://github.com/qubvel/segmentation_models/blob/master/segmentation_models/metrics.py#L25) you state only that predictions are of the shape (B, H, W, C) but you state nothing about the values in predictions. Are those probabilities (let's say `[0.2, 0.1, 0., 0.7]` for one pixel and the case with 4 class) or the one-hot encoded vectors (`[0, 0, 0, 1]` for the same pixel and the same number of classes)?

Because the result is not the same.
Let's say that the ground truth in this example is: `[0, 1, 0, 0]`.
In the first case, the intersection is 0.2, but in the second case, it is 0. And 0 is the correct intersection value.

Actually, I'm not sure how to interpret it if those are probabilities ...

**Comment by qubvel** [comment]
IoU is calculated over each class separately and than averaged. 
```
ground truth:      [0.0, 0.0, 0.0, 1.0]
prediction:        [0.2, 0.1, 0.0, 0.7] 
calss_iou_score:   [0.0, 0.0, 0.0, 0.7]  
iou_score: mean(calss_iou_score * classes_weights)
```

**Comment by qubvel** [comment]
When you train the network it is better to look at `iou_score` calculated over probabilities cause you dont know what is the best way to get thresholded values, than validate your model and find the best threshold.

**Comment by ajuric** [comment]
Ok, so you use probabilities instead of one-hot encoded predictions.

I wanted to state that you don't get the same result (same number) for IoU metric if you use probabilities for predictions and if you use one-hot encoded predictions (`1` is set to `k` where `k=argmax(prediction, axis-1)`).

**Comment by Tyler-D** [comment]
Agree with @ajuric . When training using softmax on multi-classes task (PASCAL VOC for example), I apply argmax to final feature map H*W*C to get final prediction H*W mask (each cell is the class_id ). And there is no threshold needed actually. Then I compute IOU based on the ground truth mask and prediction mask. I think it will be great if we have both kinds of IOU here.

**Comment by qubvel** [comment]
Ok, I got it. I will add this feature.
